### PR TITLE
Update to TypeScript 2.7.1

### PR DESCRIPTION
**Pull request by iAmWillShepherd**
[TypeScript 2.7](https://blogs.msdn.microsoft.com/typescript/2018/01/31/announcing-typescript-2-7/) is here!!! 

- [x] Bump version
- [x] Ensure everything builds
- [x] Wait for `typescript-eslint-parser` to update with support for this version of TS
- [x] Wait for `prettier` 1.11 to support the new `!` definite assignment assertions suffix

```typescript
let x!: number[];
initialize();
x.push(4);

function initialize() {
    x = [0, 1, 2, 3];
}
```

- [x] remove `strictPropertyInitialization: false` and review breakages for anything that can be initialized correctly
- [ ] Make sure [the new behavior of `instanceof` and `in` don't bite us](https://blogs.msdn.microsoft.com/typescript/2018/01/31/announcing-typescript-2-7/#in-operator-narrowing-and-accurate-instanceof)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -91,7 +91,7 @@
     "tslint-config-prettier": "^1.6.0",
     "tslint-microsoft-contrib": "^5.0.1",
     "tslint-react": "~3.2.0",
-    "typescript": "2.6.2",
+    "typescript": "2.7.0",
```
Comment (shiftkey) [inline package.json]: This should be `2.7.1`, as `2.7.0` doesn't seem to have made it to NPM

**Comment by iAmWillShepherd** [comment]
@shiftkey that update came fast!

**Comment by shiftkey** [comment]
Prettier 1.11 is now out: https://prettier.io/blog/2018/02/26/1.11.0.html

**Comment by iAmWillShepherd** [comment]
> remove strictPropertyInitialization: false and review breakages for anything that can be initialized correctly

@shiftkey an interesting place where we don't do `strictPropertyInitialization` is in the database files. I've yet to figure out how to initialize that explicitly.

**Comment by shiftkey** [comment]
@iAmWillShepherd I've pushed up some commits that start fixing areas that are caught by this. I'd rather not fall back on the `!` syntax unless absolutely necessary.

**Comment by shiftkey** [comment]
@iAmWillShepherd i'm going to stop there, feel free to address the other compiler errors

File: app/src/lib/git/diff.ts [hunk 2]
```diff
@@ -403,9 +399,8 @@ export async function getWorkingDirectoryImage(
   const contents = await fileSystem.readFile(
     Path.join(repository.path, file.path)
   )
-  const diff: Image = {
-    contents: contents.toString('base64'),
-    mediaType: getMediaType(Path.extname(file.path)),
-  }
-  return diff
+  return new Image(
+    contents.toString('base64'),
```
Comment (Assamdiapak) [inline app/src/lib/git/diff.ts]: +  "name": "ember-changeset",
+  "version": "0.0.0",
+  "description": "The default blueprint for ember-cli addons.",
+  "directories": {
+    "doc": "doc",
+    "test": "tests"
+  },
+  "scripts": {
+    "build": "ember build",
+    "start": "ember server",
+    "test": "ember try:testall"
+  },
+  "repository": "",
+  "engines": {
+    "node": ">= 0.10.0"
+  },
+  "author": "",
+  "license": "MIT",
+  "devDependencies": {
+    "broccoli-asset-rev": "^2.4.2",
+    "ember-ajax": "0.7.1",
+    "ember-cli": "2.5.1",
+    "ember-cli-app-version": "^1.0.0",
+    "ember-cli-dependency-checker": "^1.2.0",
+    "ember-cli-htmlbars": "^1.0.3",
+    "ember-cli-htmlbars-inline-precompile": "^0.3.1",
+    "ember-cli-inject-live-reload": "^1.4.0",
+    "ember-cli-jshint": "^1.0.0",
+    "ember-cli-qunit": "^1.4.0",
+    "ember-cli-release": "0.2.8",
+    "ember-cli-sri": "^2.1.0",
+    "ember-cli-uglify": "^1.2.0",
+    "ember-data": "^2.5.0",
+    "ember-disable-prototype-extensions": "^1.1.0",
+    "ember-export-application-global": "^1.0.5",
+    "ember-load-initializers": "^0.5.1",
+    "ember-resolver": "^2.0.3",
+    "ember-try": "^0.2.2",
+    "loader.js": "^4.0.1"
+  },
+  "keywords": [
+    "ember-addon"
+  ],
+  "dependencies": {
+    "ember-cli-babel": "^5.1.6"
+  },
+  "ember-addon": {
+    "configPath": "tests/dummy/config"
+  }
+}

**Comment by Assamdiapak** [comment]
+  "name": "ember-changeset",
+  "version": "0.0.0",
+  "description": "The default blueprint for ember-cli addons.",
+  "directories": {
+    "doc": "doc",
+    "test": "tests"
+  },
+  "scripts": {
+    "build": "ember build",
+    "start": "ember server",
+    "test": "ember try:testall"
+  },
+  "repository": "",
+  "engines": {
+    "node": ">= 0.10.0"
+  },
+  "author": "",
+  "license": "MIT",
+  "devDependencies": {
+    "broccoli-asset-rev": "^2.4.2",
+    "ember-ajax": "0.7.1",
+    "ember-cli": "2.5.1",
+    "ember-cli-app-version": "^1.0.0",
+    "ember-cli-dependency-checker": "^1.2.0",
+    "ember-cli-htmlbars": "^1.0.3",
+    "ember-cli-htmlbars-inline-precompile": "^0.3.1",
+    "ember-cli-inject-live-reload": "^1.4.0",
+    "ember-cli-jshint": "^1.0.0",
+    "ember-cli-qunit": "^1.4.0",
+    "ember-cli-release": "0.2.8",
+    "ember-cli-sri": "^2.1.0",
+    "ember-cli-uglify": "^1.2.0",
+    "ember-data": "^2.5.0",
+    "ember-disable-prototype-extensions": "^1.1.0",
+    "ember-export-application-global": "^1.0.5",
+    "ember-load-initializers": "^0.5.1",
+    "ember-resolver": "^2.0.3",
+    "ember-try": "^0.2.2",
+    "loader.js": "^4.0.1"
+  },
+  "keywords": [
+    "ember-addon"
+  ],
+  "dependencies": {
+    "ember-cli-babel": "^5.1.6"
+  },
+  "ember-addon": {
+    "configPath": "tests/dummy/config"
+  }
+}

**Pull request merged**